riscv-executor/src/profiler.rs

```diff
@@ -0,0 +1,346 @@
+use std::{
+    collections::BTreeMap,
+    fs::File,
+    io::BufWriter,
+    io::Write,
+    path::{Path, PathBuf},
+};
+
+use itertools::Itertools;
+
+use rustc_demangle::demangle;
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Call<'a> {
+    from: Loc<'a>,
+    target: Loc<'a>,
+}
+
+/// Risc-v asm profiler.
+/// Tracks the self-cost of functions and cummulative cost specific function calls (i.e., callgrind style).
+pub struct Profiler<'a> {
+    /// profiling options
+    options: ProfilerOptions,
+    /// file number to (dir,file)
+    debug_files: &'a [(&'a str, &'a str)],
+    /// pc value of function beginnings
+    function_begin: BTreeMap<usize, &'a str>,
+    /// pc value of .debug loc statements
+    location_begin: BTreeMap<usize, (usize, usize)>,
+    /// current call stack, entries include running cost
+    call_stack: Vec<(Call<'a>, usize)>,
+    /// saved return address of "jump and link" instructions
+    return_pc_stack: Vec<usize>,
+    /// cost of each location
+    location_stats: BTreeMap<Loc<'a>, usize>,
+    /// (count, cummulative cost) of calls
+    call_stats: BTreeMap<Call<'a>, (usize, usize)>,
+    /// stack sampling format for FlameGraph
+    folded_stack_stats: BTreeMap<Vec<&'a str>, usize>,
+}
+
+#[derive(Default, Clone)]
+pub struct ProfilerOptions {
+    pub output_directory: String,
```

Comment:
In other crates we try to keep FS access to a minimal, and rather return results that CLI then writes to the FS. I guess we could do the same here without really disrupting the architecture?

Reply:
or you could take an `impl Write` here instead of the file path